lib/chef/resource/inspec_waiver_file.rb

```diff
@@ -0,0 +1,153 @@
+          action :add
+        end
+      ```
+
+      **Add an InSpec waiver entry to a given waiver file using the 'name' property to identify the control**:
+
+      ```ruby
+        inspec_waiver_file 'my_inspec_control_01' do
+          file 'C:\\chef\\inspec_waiver.yml'
+          justification "The subject of this control is not managed by Chef on the systems in policy group \#{node['policy_group']}"
+          action :add
+        end
+      ```
+
+      **Remove an InSpec waiver entry to a given waiver file**:
+
+      ```ruby
+        inspec_waiver_file "my_inspec_control_01" do
+          file '/etc/chef/inspec_waiver.yml'
+          action :remove
+        end
+      ```
+      DOC
+
+      property :control, String,
+        name_property: true,
+        description: "The name of the control being added or removed to the waiver file"
+
+      property :file, String,
+        required: true,
+        description: "The path to the waiver file being modified"
+
+      property :expiration, String,
+        description: "The expiration date of the given waiver - provided in YYYY-MM-DD format",
+        callbacks: {
+          "Expiration date should be a valid calendar date and match the following format: YYYY-MM-DD" => proc { |e|
+            re = Regexp.new('\d{4}-\d{2}-\d{2}$').freeze
+            if re.match?(e)
+              Date.valid_date?(*e.split("-").map(&:to_i))
+            elsif e.nil?
+              true
+            else
+              false
+            end
+          },
+        }
+
+      property :run_test, [true, false],
+        description: "If present and true, the control will run and be reported, but failures in it won’t make the overall run fail. If absent or false, the control will not be run."
+
+      property :justification, String,
+        description: "Can be any text you want and might include a reason for the waiver as well as who signed off on the waiver."
+
+      property :backup, [false, Integer],
+        description: "The number of backups to be kept in /var/chef/backup (for UNIX- and Linux-based platforms) or C:/chef/backup (for the Microsoft Windows platform). Set to false to prevent backups from being kept.",
+        default: false
+
+      action :add do
+        filename = new_resource.file
+        waiver_hash = load_waiver_file_to_hash(filename)
+        control_hash = {}
+        control_hash["expiration_date"] = new_resource.expiration.to_s unless new_resource.expiration.nil?
+        control_hash["run"] = new_resource.run_test unless new_resource.run_test.nil?
+        control_hash["justification"] = new_resource.justification.to_s
+
+        if waiver_hash.key?(new_resource.control)
+          unless waiver_hash[new_resource.control] == control_hash
+            waiver_hash.delete(new_resource.control)
+            waiver_hash[new_resource.control] = control_hash
+          end
+        else
+          waiver_hash[new_resource.control] = control_hash
+        end
```

Comment:
```suggestion
        waiver_hash[new_resource.control] = control_hash
```

If this code is making sure that `waiver_hash` has the new `control_hash` set for the key `new_resource.control`, can it just always set that value and avoid all of this logic? I walked through this a few times and it seems like this should be equivalent.

Reply:
so, it was validating that there were no changes to the expiration, run_tests, or justification of a control and then not bothering to run the file resource.  but I guess we could always just wipe out the control from the hash, recreate it and then let the file resource determine if there's a diff or not...

Reply:
I was curious if that's what was going on, and in that case you could wrap the file resource in a `if waiver_hash[new_resource_control] == control_hash`.